src/python/pants/backend/python/typecheck/mypy/rules.py

```diff
@@ -275,16 +297,81 @@ async def mypy_typecheck(
         FallibleProcessResult,
         PexProcess(
             mypy_pex,
-            argv=generate_args(mypy, file_list_path=file_list_path),
+            argv=generate_argv(mypy, file_list_path=file_list_path, python_version=python_version),
             input_digest=merged_input_files,
             extra_env=env,
             description=f"Run MyPy on {pluralize(len(typechecked_srcs_snapshot.files), 'file')}.",
             level=LogLevel.DEBUG,
         ),
     )
-    return TypecheckResults(
-        [TypecheckResult.from_fallible_process_result(result)], typechecker_name="MyPy"
+    return TypecheckResult.from_fallible_process_result(
+        result, partition_description=str(sorted(str(c) for c in partition.interpreter_constraints))
+    )
+
+
+# TODO(#10131): Improve performance, e.g. by leveraging the MyPy cache.
+# TODO(#10131): Support .pyi files.
+@rule(desc="Typecheck using MyPy", level=LogLevel.DEBUG)
+async def mypy_typecheck(
+    request: MyPyRequest, mypy: MyPy, python_setup: PythonSetup
+) -> TypecheckResults:
+    if mypy.skip:
+        return TypecheckResults([], typechecker_name="MyPy")
+
+    # We batch targets by their interpreter constraints to ensure, for example, that all Python 2
+    # targets run together and all Python 3 targets run together. We can only do this by setting
+    # the `--python-version` option, but we allow the user to set it as a safety valve. We warn if
+    # they've set the option.
+    config_content = await Get(DigestContents, PathGlobs, config_path_globs(mypy))
+    python_version_configured = check_and_warn_if_python_version_configured(
+        config=next(iter(config_content), None), args=mypy.args
+    )
+
+    # When determining how to batch by interpreter constraints, we must consider the entire
+    # transitive closure to get the final resulting constraints.
+    transitive_targets_per_field_set = await MultiGet(
+        Get(TransitiveTargets, Addresses([field_set.address])) for field_set in request.field_sets
+    )
```

Comment:
This is my main concern. It adds about 1.5 seconds to running MyPy on the Pants codebase. Performance is already an issue with Pants and MyPy.

I don't know if we should feature-gate partitioning feature.

Reply:
Does this `await MultiGet` cause the 1.5 second slowdown, or the corresponding additional runs of mypy? I would expect that when running mypy over the *entire* pants codebase, that 1.5 seconds more wouldn't be too much to ask?

Reply:
My benchmark was running this specific `MulitGet` one-fieldset-at-a-time vs. including all the field sets in the input `Addresses` for one single `Get`. I used `time.time()` to measure.

Idrk what the right answer is with feature gating. I think for sure we want to land this feature, as it's a really helpful capability. My concern with feature gating is that you might turn it off without realizing the need. For example, Toolchain used to be `>=3.8` always. Recently, we made a portion of our codebase `>=3.7`. While it's not the end of the world if we use `>=3.8` on that too, we actually do want the partitioning imo because it will mean MyPy can check that we're not using Py38 syntax on accident. I don't think we would have realized we should turn on this partitioning when making the change; it's subtle.

Feature gating also makes this code more complex, although that's not something we should optimize too much for imo.

On the other hand, 1.5 seconds when Pants is already super slow with MyPy isn't great.

Reply:
> it will mean MyPy can check that we're not using Py38 syntax on accident

Isn't that going to be checked just by running the code as well? It'd be much more annoying, obviously, but it wouldn't be a horribly silent error, I think.

Reply:
> Isn't that going to be checked just by running the code as well

Only if we actually run the code with Py37. For that particular code, we don't have many tests yet. And we don't yet do what is really robust: creating Py37 _and_ Py38 targets for the same test file to ensure it passes with both interpreters. (Or, doing what Pants does to run nightly CI by changing the global default). 

So, MyPy running with Py37 is a simple, automated thing that would catch issues in the meantime. Along with catching type issues that tests might not catch if they aren't comprehensive enough.

Reply:
Ok, that makes sense.

Reply:
Given the above, I would personally opt for taking the 1.5 second hit to avoid silent errors (especially since feature gating would directly affect toolchain). I think there are likely to be other places to improve performance, after profiling, if that becomes even mildly annoying.

Reply:
I bias towards the correctness argument, too. You don't need partitioning Until You Do.

I think Stu had some ideas for how we might be able to make the calls to `TransitiveTargets` faster. And I didn't benchmark using pantsd, so it might be less offensive there.

But still, 1.5 seconds is substantial when the purpose of our tool is to speed up your workflow.

Reply:
> But still, 1.5 seconds is substantial when the purpose of our tool is to speed up your workflow.

Avoiding the need to manually construct the mypy command line is a reduction in mental energy which (along with not having to construct the same for pytest, etc) would be the tradeoff we can intentionally offer (*if* we find mypy is stubbornly difficult to speed up).

Reply:
Yes, I think that we should land this and partition by default. Allowing for "compute the transitive targets for all of these roots" should be a relatively simple API change to TT to preserve the graph structure in the result, and then expose methods to get the closure on a root-by-root basis.

Reply:
Adding a TODO here, and a few other comments in the body of this method would be good though.